### DISCUSSION TITLE
instancing

### DISCUSSION THREAD

**Discussion by unicomp21** [google/filament#4132] [General]
https://github.com/google/filament/issues/1056#issuecomment-645554619

^how is this done?

**Comment by jnmaloney** [comment]
I think it's to create a `VertexBuffer`  `IndexBuffer` and then make repeated calls to `RenderableManager::Builder::geometry`.

**Reply by unicomp21** [reply]
Is there something similar to per-instance data within a renderable?  Like have multiple, different attributes such as scale, position?  Where these get applied to the same VertexBuffer,IndexBuffer?  If so, can these per-instance attributes be updated in realtime?

**Reply by jnmaloney** [reply]
Something like this. Each Entity is an effectively an instance, but each one has to be built as a renderable. 

```  
  utils::Entity entity;  
  auto& tcm = engine->getTransformManager();  
  tcm.setTransform(  
    tcm.getInstance(entity),   
    transform  
  );
```

**Reply by unicomp21** [reply]
thanks!

**Reply by unicomp21** [reply]
is there an example, kinda like a particle system?  with a large number of instances?

**Reply by prideout** [reply]
As an aside, note that a particle system can be implemented with points rather than instancing, which we have an example of (`point_sprites.cpp`)

If you cannot use point sprites and you have a huge number of instances, you can consider using a vertex texture for storing pivot points, I have an example of this [here](https://github.com/google/filament/commit/bb6c22bb5976279923475e5856afdb705a28b566), although it uses our JavaScript API.

**Reply by unicomp21** [reply]
I'm trying to do something like this, am I on the right track? @prideout 

https://playground.babylonjs.com/#GLZ1PX#9
https://doc.babylonjs.com/divingDeeper/particles/solid_particle_system/sps_animate
